internal/webhooks/clusterclass.go

```diff
@@ -123,8 +142,88 @@ func (webhook *ClusterClass) validate(old, new *clusterv1.ClusterClass) error {
+	allErrs = append(allErrs, webhook.validateRemovedMachineDeploymentClassesAreNotUsed(old, new)...)
+
 	if len(allErrs) > 0 {
 		return apierrors.NewInvalid(clusterv1.GroupVersion.WithKind("ClusterClass").GroupKind(), new.Name, allErrs)
 	}
 	return nil
 }
+
+func (webhook *ClusterClass) validateRemovedMachineDeploymentClassesAreNotUsed(old, new *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+	if old == nil {
+		return allErrs
+	}
+	removedClasses := webhook.removedMachineClasses(old, new)
+
+	// If no classes have been removed return early as no further checks are needed.
+	if len(removedClasses) == 0 {
+		return allErrs
+	}
+
+	// Retrieve all clusters using the ClusterClass.
+	clusters, errs := webhook.getClustersUsingClusterClass(old)
+	if len(errs) > 0 {
+		allErrs = append(allErrs, errs...)
+		return allErrs
+	}
+
+	// Error if any Cluster using the ClusterClass uses a MachineDeploymentClass that has been removed.
+	for _, c := range clusters {
+		for i, machineDeploymentTopology := range c.Spec.Topology.Workers.MachineDeployments {
+			if removedClasses.Has(machineDeploymentTopology.Class) {
+				// TODO(killianmuldoon): Improve error printing here so large scale changes don't flood the error log e.g. deduplication, only example usages given.
+				allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "topology", "workers", "machineDeployments").Index(i).Child("class"),
+					fmt.Sprintf("MachineDeploymentClass %v is in use in MachineDeploymentTopology %v in Cluster %v. ClusterClass %v modification not allowed",
+						machineDeploymentTopology.Class, machineDeploymentTopology.Name, c.Name, old.Name),
+				))
+			}
+		}
+	}
+	return allErrs
+}
+
+func (webhook *ClusterClass) removedMachineClasses(old, new *clusterv1.ClusterClass) sets.String {
+	removedClasses := sets.NewString()
+
+	classes := webhook.classNamesFromWorkerClass(new.Spec.Workers)
+	for _, oldClass := range old.Spec.Workers.MachineDeployments {
+		if !classes.Has(oldClass.Class) {
+			removedClasses.Insert(oldClass.Class)
+		}
+	}
+	return removedClasses
+}
+
+// classNamesFromWorkerClass returns the set of MachineDeployment class names.
+func (webhook *ClusterClass) classNamesFromWorkerClass(w clusterv1.WorkersClass) sets.String {
+	classes := sets.NewString()
+	for _, class := range w.MachineDeployments {
+		classes.Insert(class.Class)
+	}
+	return classes
+}
+
+func (webhook *ClusterClass) getClustersUsingClusterClass(clusterClass *clusterv1.ClusterClass) ([]clusterv1.Cluster, field.ErrorList) {
+	var allErrs field.ErrorList
+	clusters := &clusterv1.ClusterList{}
+	clustersUsingClusterClass := []clusterv1.Cluster{}
+	err := webhook.Client.List(context.Background(), clusters,
+		client.MatchingLabels{
+			clusterv1.ClusterTopologyOwnedLabel: "",
+		},
+		client.InNamespace(clusterClass.Namespace),
+	)
+	if err != nil {
+		allErrs = append(allErrs, field.InternalError(field.NewPath(""), err))
```

Comment:
I think it would be good to wrap the error.

I wonder if there is a different way to express this error, given that it's not related to a field of the resource we currently validate (ClusterClass) and thus not really a field error. Not sure how this user currently looks like user-facing though. 
Maybe we should make this an apierror instead. (I think it's doable if we move the `getClustersUsingClusterClass` call one layer up to `validate`)

Reply:
+1 to make this an apierror